docs/source/traits_user_manual/internals.rst

```diff
@@ -0,0 +1,190 @@
+setting the corresponding attribute, as well as providing an attachment point
+for trait notifications.
+
+In addition to these two dictionaries, a |HasTraits| object has a ``__dict__``
+that stores attribute values for that object in the normal way.
+
+Analogously to class variables and instance variables in a normal Python class,
+the class traits dictionary for a given |HasTraits| subclass is shared between
+instances of that class, while the instance traits dictionary is specific to a
+particular |HasTraits| instance. (As an internal optimization, the instance
+traits dictionary is created on demand when first needed, while the class
+traits dictionary is always present for each instance.)
+
+For introspection and debugging purposes, both of these dictionaries can
+be retrieved directly, using the |_instance_traits| and |_class_traits|
+methods. However, it's not recommended to use either of these methods in
+production code, and you should be especially careful when modifying either
+of these dictionaries.
+
+For example, consider the following |HasTraits| subclass::
+
+    from traits.api import Float, HasTraits, Str
+
+    class Ingredient(HasTraits):
+        """ An ingredient in a recipe listing. """
+
+        #: The name of the ingredient.
+        name = Str()
+
+        #: Ingredient quantity.
+        quantity = Float()
+
+Here's what happens when we create an instance of this ingredient and inspect
+the instance and class trait dictionaries::
+
+    >>> eggs = Ingredient(name="eggs", quantity=12)
+    >>> eggs._instance_traits()
+    {}
+    >>> eggs._class_traits()
+    {
+        'name': <traits.ctrait.CTrait object at 0x1020cd400>,
+        'quantity': <traits.ctrait.CTrait object at 0x1020cd360>,
+        'trait_added': <traits.ctrait.CTrait object at 0x1020cd4a0>,
+        'trait_modified': <traits.ctrait.CTrait object at 0x1020079a0>,
+    }
+    >>> eggs.__dict__
+    {'name': 'eggs', 'quantity': 12.0}
+
+Note that the actual values for the ingredient are stored in the ``__dict__``
+as usual, not in the ``CTrait`` objects.
+
+If we create a second ingredient, it shares class traits (but not
+instance traits) with the first one::
+
+    >>> flour = Ingredient(name="flour", quantity=3.5)
+    >>> flour._class_traits() is eggs._class_traits()
+    True
+    >>> flour._instance_traits() is eggs._instance_traits()
+    False
+
+
+Key class: |CTrait|
+~~~~~~~~~~~~~~~~~~~
+
+A |CTrait| object has two main purposes:
+
+- It encapsulates the rules for getting and setting a traited attribute on
+  a |HasTraits| object.
+- It provides an attachment point for trait notifiers.
+
+For attribute access, the C structure underlying each |CTrait| instance (see
+the ``trait_object`` struct in the C source) has three relevant fields:
+``getattr``, ``setattr`` and ``post_setattr``. These fields contain C function
+pointers, and are invoked during attribute get and set operations, as described
+below.
```

Comment:
Just to check my understanding: I think the "C structure underlying each `CTrait` instance" is referring to the object referenced by `CTrait.handler`?

Reply:
Not the handler; I meant the `trait_object` C struct. I'll attempt to reword to clarify.

Reply:
That is, I'm referring to these three struct fields: https://github.com/enthought/traits/blob/de2eaa29e521d688e0e97935ccc5c98e827b0201/traits/ctraits.c#L217-L219

Reply:
I've reworded the paragraph, though I think what's really needed is a much fuller description of the mechanics. Any better?